[UnitTests] Require cached fixtures to be copy-able, with opt-in.

Lunderberg opened the pull request:
Previously, any class that doesn't raise a TypeError in copy.deepcopy could be used as a return value in a `@tvm.testing.fixture`.  This has the possibility of incorrectly copying classes inherit the default `object.__reduce__` implementation.  Therefore, only classes that explicitly implement copy functionality (e.g. __deepcopy__ or __getstate__/__setstate__), or that are explicitly listed in tvm.testing._fixture_cache are allowed to be cached.

Lunderberg commented:
@areusch This is an implementation of Option 3 from [the discussion](https://github.com/apache/tvm-rfcs/pull/7#issuecomment-875834029) on the parametrized unit test RFC.

emijiayw commented on python/tvm/testing.py:
```diff
@@ -1160,6 +1162,43 @@ def wraps(func):
+                "to be copied (e.g. through __deepcopy__, __getstate__, or __setstate__),"
+                "and forbids the use of the default `object.__reduce__` and "
+                "`object.__reduce_ex__`.  For third-party classes that are known to be "
+                "safe to use with copy.deepcopy, please add the class to "
```
For third-party classes that are safe to use copy.deepcopy, please add these classes into

Lunderberg replied:
Good call, the "known to be" is unnecessarily wordy.  Edited as suggested.

emijiayw commented on python/tvm/testing.py:
```diff
@@ -1199,18 +1238,15 @@ def wrapper(*args, **kwargs):
+                cached_value,
+                _DeepCopyAllowedClasses(
+                    # *args should be a list of classes that are known
+                    # to be safe to copy using copy.deepcopy, but do
```
that are safe to copy with copy.deepcopy

Lunderberg replied:
Agreed as above.

areusch reviewed (commented):
thanks @Lunderberg just a couple questions!

Lunderberg commented on python/tvm/testing.py:
```diff
@@ -1160,6 +1162,43 @@ def wraps(func):
+        )
+        raise TypeError(
+            (
+                "Cannot copy fixture of type {}.  TVM fixture caching "
```
Sure.  Somehow I had it in my mind that f-strings were added in 3.7, not 3.6, and so I was avoiding them since TVM targets 3.6.

Lunderberg commented on python/tvm/testing.py:
```diff
@@ -1160,6 +1162,43 @@ def wraps(func):
+        self.allowed_class_list = allowed_class_list
+        super().__init__()
+
+    def get(self, key, *args, **kwargs):
```
And documented.

Lunderberg commented on python/tvm/testing.py:
```diff
@@ -1160,6 +1162,43 @@ def wraps(func):
 
 
+class _DeepCopyAllowedClasses(dict):
+    def __init__(self, *allowed_class_list):
```
Mostly that it looked a bit cleaner to me at the call-site (entirely subjective).  I'm fine with it either way, and have switched it to be a non-`*args` parameter instead.

Lunderberg commented:
@areusch Any additional concerns before merging?

Lunderberg commented:
Rebased onto main to resolve git merge conflict, no issues expected in CI.

areusch commented on tests/python/unittest/test_tvm_testing_features.py:
```diff
@@ -210,5 +210,44 @@ def test_pytest_mark_covariant(self, request, target, other_param):
+    def uncacheable_fixture(self):
+        return self.EmptyClass()
+
+    @pytest.mark.xfail(reason="Requests cached fixture of uncacheable type", strict=True)
```
is there any way to assert on the specific error generated here?

Lunderberg replied:
Normally, yes, with `pytest.raises`.  Unfortunately, I don't think that's possible here, because the exception is raised during the fixture function definition.

areusch replied:
seems like you could just invoke uncacheable_fixture from the test body, no?

Lunderberg replied:
Good point.  It can't be called directly, but `request.getfixturevalue('uncacheable')` does trigger it within the test.  Updated.

areusch commented on tests/python/unittest/test_tvm_testing_features.py:
```diff
@@ -210,5 +210,44 @@ def test_pytest_mark_covariant(self, request, target, other_param):
+    bool(int(os.environ.get("TVM_TEST_DISABLE_CACHE", "0"))),
+    reason="Cannot test cache behavior while caching is disabled",
+)
+class TestCacheableTypes:
```
is there any way to assert you're getting a copy in each test case?

Lunderberg replied:
Good call, and I've added a `TestCachedFixtureIsCopy` to verify that the cached fixture received is a copy.

areusch reviewed (commented):
thanks @Lunderberg , i think this looks pretty good now. two more questions about the tests, then i think we can merge.

Lunderberg commented:
Updated based on comments, and rebased on main.  The `TestAutoParametrization` caught an issue with cudnn parametrization (added to default targets in #8383), which is now also resolved in this PR.

Lunderberg commented:
Latest CI failure is due to flaky `test_conv2d_nchw`.  Will restart CI after #8723 gets merged.

Lunderberg commented:
After the post-flaky-test CI restart, CI has passed.  Any last changes, @areusch ?

areusch reviewed (approved):
thanks @Lunderberg !
